File: app/src/main/java/org/dhis2/data/service/SyncGranularRxModule.java
package org.dhis2.data.service;

import androidx.annotation.NonNull;

import org.dhis2.data.dagger.PerService;
import org.dhis2.data.prefs.PreferenceProvider;
import org.dhis2.data.service.workManager.WorkManagerController;
import org.dhis2.utils.analytics.AnalyticsHelper;
import org.hisp.dhis.android.core.D2;

import dagger.Module;
import dagger.Provides;

@Module
@PerService
public class SyncGranularRxModule {

    @Provides
    @PerService
    SyncPresenter syncPresenter(
            @NonNull D2 d2,
            @NonNull PreferenceProvider preferences,
            @NonNull WorkManagerController workManagerController,
            @NonNull AnalyticsHelper analyticsHelper
    ) {
        return new SyncPresenterImpl(d2, preferences, workManagerController, analyticsHelper);
    }
}

File: app/src/main/java/org/dhis2/data/service/SyncGranularRxModule.kt
package org.dhis2.data.service

import dagger.Module
import dagger.Provides
import org.dhis2.commons.di.dagger.PerService
import org.dhis2.commons.prefs.PreferenceProvider
import org.dhis2.mobile.sync.domain.SyncStatusController
import org.hisp.dhis.android.core.D2

@Module
class SyncGranularRxModule(
    private val syncStatusController: SyncStatusController,
) {
    @Provides
    @PerService
    fun syncRepository(d2: D2): SyncRepository = SyncRepositoryImpl(d2)

    @Provides
    @PerService
    internal fun syncPresenter(
        d2: D2,
        preferences: PreferenceProvider,
        syncRepository: SyncRepository,
    ): SyncPresenter =
        SyncPresenterImpl(
            d2,
            preferences,
            syncRepository,
            syncStatusController,
        )
}
